src/main/java/org/broadinstitute/hellbender/utils/pairhmm/LazyLoadingReferenceNucleotideSequence.java

```diff
@@ -0,0 +1,79 @@
+package org.broadinstitute.hellbender.utils.pairhmm;
+
+import htsjdk.samtools.SAMSequenceRecord;
+import org.broadinstitute.hellbender.engine.ReferenceDataSource;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+
+import java.util.concurrent.*;
+
+public class LazyLoadingReferenceNucleotideSequence implements NucleotideSequence {
```

Comment:
javadoc

Reply:
Looking at the places that you use this class (just in traverse() of `SampleSitesForDRAGstrModel` I don't think you are actually gaining very much by making this a lazy loader that sits on top of the reference datasource. As its written you are almost never saving yourself the loads. Indeed it looks like for each site you are instantiating a new one of these objects and immediately filling the first 10000 bases from the reference data source. There is already a cache in the reference data source that is designed to handle exactly this use case (traversing over the whole reference and grabbing bases ~in order).